# Analyzing borrowers’ risk of defaulting

Our project is to prepare a report for a bank’s loan division. We will need to find out if a customer’s marital status and number of children has an impact on whether they will default on a loan. The bank already has some data on customers’ credit worthiness.

Our report will be considered when building a **credit scoring** of a potential customer. A **credit scoring** is used to evaluate the ability of a potential borrower to repay their loan.

## Open the data file and have a look at the general information. 

In [5]:
#import pandas library.
import pandas as pd
#make a try-except block in order to avoid path error.
try:
    df = pd.read_csv('credit_scoring_eng.csv')
except:
    df = pd.read_csv('/datasets/credit_scoring_eng.csv')
    
#check the table's first 10 rows to see the general information.     
df.head(10)

children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   
5         0    -926.185831         27    bachelor's degree             0   
6         0   -2879.202052         43    bachelor's degree             0   
7         0    -152.779569         50  SECONDARY EDUCATION             1   
8         2   -6929.865299         35    BACHELOR'S DEGREE             0   
9         0   -2188.756445         41  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   
5  civil partnership                 1      M    business     0     40922.170   
6            married                 0      F    business     0     38484.156   
7            married                 0      M    employee     0     21731.829   
8  civil partnership                 1      F    employee     0     15337.093   
9            married                 0      M    employee     0     23108.150   

                               purpose  
0                purchase of the house  
1                         car purchase  
2                purchase of the house  
3              supplementary education  
4                    to have a wedding  
5                purchase of the house  
6                 housing transactions  
7                            education  
8                     having a wedding  
9  purchase of the house for my family

In [2]:
#Check the table's general structure.
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [3]:
#Check the table's general statistics.
df.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt   total_income  
count      21525.000000  21525.000000   19351.000000  
mean           0.972544      0.080883   26787.568355  
std            1.420324      0.272661   16475.450632  
min            0.000000      0.000000    3306.762000  
25%            0.000000      0.000000   16488.504500  
50%            0.000000      0.000000   23202.870000  
75%            1.000000      0.000000   32549.611000  
max            4.000000      1.000000  362496.645000

In [4]:
#Show the missing values that the .info() method showed us more clear.
df.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

In [5]:
#how much missing values comapred to total values in column
null_prec = df[['days_employed','total_income']].isnull().sum() / len(df)
null_prec

days_employed    0.100999
total_income     0.100999
dtype: float64

In [22]:
df[['days_employed','total_income']].isna().mean()

days_employed    0.0
total_income     0.0
dtype: float64

In [6]:
#print the rows with duplicates - 
has_duplicate = df.duplicated()
duplicates = df[has_duplicate]
duplicates

children  days_employed  dob_years            education  education_id  \
2849          0            NaN         41  secondary education             1   
4182          1            NaN         34    BACHELOR'S DEGREE             0   
4851          0            NaN         60  secondary education             1   
5557          0            NaN         58  secondary education             1   
7808          0            NaN         57  secondary education             1   
8583          0            NaN         58    bachelor's degree             0   
9238          2            NaN         34  secondary education             1   
9528          0            NaN         66  secondary education             1   
9627          0            NaN         56  secondary education             1   
10462         0            NaN         62  secondary education             1   
10697         0            NaN         40  secondary education             1   
10864         0            NaN         62  secondary education             1   
10994         0            NaN         62  secondary education             1   
11791         0            NaN         47  secondary education             1   
12373         0            NaN         58  secondary education             1   
12375         1            NaN         37  secondary education             1   
12736         0            NaN         59  secondary education             1   
13025         1            NaN         44  secondary education             1   
13639         0            NaN         64  secondary education             1   
13773         0            NaN         35  secondary education             1   
13878         1            NaN         31  secondary education             1   
13942         0            NaN         44  secondary education             1   
14432         2            NaN         36    bachelor's degree             0   
14832         0            NaN         50  secondary education             1   
15091         0            NaN         58  secondary education             1   
15188         0            NaN         60  secondary education             1   
15273         0            NaN         57  secondary education             1   
16176         0            NaN         47  secondary education             1   
16378         0            NaN         46  secondary education             1   
16902         2            NaN         39  secondary education             1   
16904         1            NaN         32    bachelor's degree             0   
17379         0            NaN         54    bachelor's degree             0   
17755         1            NaN         43  secondary education             1   
17774         1            NaN         40  secondary education             1   
18328         0            NaN         29    bachelor's degree             0   
18349         1            NaN         30    bachelor's degree             0   
18428         0            NaN         64  secondary education             1   
18521         0            NaN         56  secondary education             1   
18563         0            NaN         54  secondary education             1   
18755         0            NaN         58  secondary education             1   
19041         0            NaN         56  secondary education             1   
19184         0            NaN         46  secondary education             1   
19321         0            NaN         23  secondary education             1   
19387         0            NaN         38    bachelor's degree             0   
19688         0            NaN         61  secondary education             1   
19832         0            NaN         48  secondary education             1   
19946         0            NaN         57  secondary education             1   
20116         0            NaN         57  secondary education             1   
20165         0            NaN         42  secondary education             1   
20702  

In [7]:
#Number of duplicated rows
df.duplicated().sum()

54

In [8]:
#checking for the count of those unique values
df['gender'].value_counts()

F      14236
M       7288
XNA        1
Name: gender, dtype: int64

### Conclusion

#### After having a look at the table's general information, We discovered useful information regarding our data.

#### 1. There are 2174 rows with 2 columns that contain missing values (days_employed, total_income).
My assumption is that the fact both columns are missing together in every row, means the customer is not working,
Having days_employed and total_income missing together makes it easy to belive the person is not employed(doesnt have days Employed and no income). I would ask the provider of the data if that's his intention by leaving those columns Missing together.
The precentage of unemployed persons by this logic is 10% roughly (10.099%).
Those missing values will be filled with 0's accordingly due to the logic in this scenario. (0 days employed and 0 income when not working).
#### 2. There are data types that need to be changed.
days_employed will be changed to int as I didnt find this information useful as a float.
all int type columns will be downcasted to the lowest possible type in order to save memory.
#### 3. There are illogical values in the table.
Thanks to the describe() method, we are able to see some illogical values in some of the columns.
a) Value of -1 children is not possible, and value of 20 children can be possible if the amount families is relatively low (will be checked).
b) Negative value of days_employed cant be possible, we can assume those -'s are filled in by a mistake and take the absolute value of those rows with negative values. I would ask the data provider about it.
There are also illogical values in days_employed, such as 401755 days employed (1,100 years), we will take care of this by making a logical range of employed days.
c)age of 0 in dob_years is not possible, will be changed to a logical value of 18 as minimum age for loan request.
d)gender contains 1 row with value 'XNA'
#### 4. There are duplicated rows.
There are 54 duplicated rows that will be removed.
Used the duplicated.sum() methods in order to see the amount of duplicated rows.


## Data preprocessing

### Processing missing values

In [9]:
#replacing missing values (only exist in days_employed and total_income) with 0.
df['days_employed'].fillna(0, inplace = True)
df['total_income'].fillna(0, inplace = True)

#make sure the missing values are gone and replaced with 0.
df.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Conclusion

#### 
we have replaced the missing values with 0 using the .fillna method.
We made sure there are no missing values left in our data with .isna.sum 

### Proccecing illogical values

In [10]:
#Considering -1 value for children as human error, we will take the absolute value.
df[df['children'] == -1].count()

#seeing that we only have 47 children with -1 value, we realize it won't have an impact on the data.
#Using the .abs method to get the absolute values in the children column.
df['children'] = df['children'].abs()

#Checking all values of children.
df['children'].value_counts()

#to have value of 20 children appear 76 times is highly unlikely, and we can consider it as human error that accidently entered 
#20 instead of 2, we will turn the 20 values to 2 using the .replace method.
df['children'].replace(to_replace =20,value =2, inplace=True)

#checking that the replacement worked.
df['children'].value_counts()

#checking for ages below 18  in the dob_years column
df[df['dob_years'] <= 18]

#there are 101 rows with 0 value, we will remove them due to the small amount and lack of impact on the data.
df.drop(df[df['dob_years'] == 0].index, inplace=True)

#Using the .abs method to get the absolute values in the days_employed columns.
df['days_employed'] = df['days_employed'].abs()

#checking the minimum value to confirm no negative values exist.
df['days_employed'].min()

#Checking for days_employed > dob_years -18 values which are highly unlikely - them starting to work before the age of 18,
#making a variable 'high_days' to store those rows.
high_days = df[df['days_employed'] / 365 > (df['dob_years'] - 18)]
high_days

#There are 3835 rows where this condition applies, Which their 'days_employed' value will be replaced with the a new value
#Of 80% of their age starting from age 18, formula:  365 * ((dob_years - 18) * 0.8).
high_days['days_employed'] = 365 * ((high_days['dob_years'] - 18)* 0.8)

#replacing those rows in the original dataframe.
df.loc[df['days_employed'] / 365 > (df['dob_years'] - 18) , 'days_employed'] = high_days['days_employed']

#adding years_employed column for better visualization.
df['years_employed'] = df['days_employed'] / 365

#remove the row with 'XNA' value of gender.
df.drop(df[df['gender'] == 'XNA'].index, inplace=True)

df.describe()

/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:37: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy


children  days_employed     dob_years  education_id  \
count  21423.000000   21423.000000  21423.000000  21423.000000   
mean       0.479671    3644.289940     43.498390      0.817859   
std        0.755304    4272.682630     12.246495      0.548178   
min        0.000000       0.000000     19.000000      0.000000   
25%        0.000000     603.325492     33.000000      1.000000   
50%        0.000000    1799.549893     43.000000      1.000000   
75%        1.000000    4709.018460     53.000000      1.000000   
max        5.000000   16352.000000     75.000000      4.000000   

       family_status_id          debt   total_income  years_employed  
count      21423.000000  21423.000000   21423.000000    21423.000000  
mean           0.971293      0.080894   24087.579227        9.984356  
std            1.419770      0.272679   17600.501548       11.705980  
min            0.000000      0.000000       0.000000        0.000000  
25%            0.000000      0.000000   14179.846000        1.652947  
50%            0.000000      0.000000   21683.559000        4.930274  
75%            1.000000      0.000000   31278.989000       12.901420  
max            4.000000      1.000000  362496.645000       44.800000

In [11]:
#df2 = df[["children", "days_employed", "years_employed",'education','education_id','family_status','family_status_id','gender','income_type','debt','total_income','income_level','purpose','purpose_type']]
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21423 entries, 0 to 21524
Data columns (total 13 columns):
children            21423 non-null int64
days_employed       21423 non-null float64
dob_years           21423 non-null int64
education           21423 non-null object
education_id        21423 non-null int64
family_status       21423 non-null object
family_status_id    21423 non-null int64
gender              21423 non-null object
income_type         21423 non-null object
debt                21423 non-null int64
total_income        21423 non-null float64
purpose             21423 non-null object
years_employed      21423 non-null float64
dtypes: float64(3), int64(5), object(5)
memory usage: 2.3+ MB


#### Conclusion
Changed the children and days_employed values to absolute values (.abs() ) in order to get rid of the negative values.
Assumed value of 20 children is too high and is meant to be 2, so we made the replacement.
Dropped the rows with dob_years = 0 due to the small number of rows.
Made the high_days variable to store the problematic rows (people who started to work before the age of 18),
And changed the values of days_employed to 80% of their age -18 (assuming the starting working age is 18).
Replaced those rows in the original dataframe.
Removed row of'XNA' gender value.

### Data type replacement

In [12]:
#replacing the datatype for the 'employed_days' column from object to int as we dont need this info as a float.
df['days_employed']= df['days_employed'].astype('int')

#downcasting int type columns to lowest possible int type.
df['days_employed'] = pd.to_numeric((df['days_employed']), downcast ='integer')
df['children'] = pd.to_numeric((df['children']), downcast ='integer')
df['dob_years'] = pd.to_numeric((df['dob_years']), downcast ='integer')
df['education_id'] = pd.to_numeric((df['education_id']), downcast ='integer')
df['family_status_id'] = pd.to_numeric((df['family_status_id']), downcast ='integer')
df['debt'] = pd.to_numeric((df['debt']), downcast ='integer')

#changing total_income float type to lowest possible float type.
df['total_income'] = pd.to_numeric((df['total_income']), downcast ='float')
#checking to see if the integer type managed to downcast
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21423 entries, 0 to 21524
Data columns (total 13 columns):
children            21423 non-null int8
days_employed       21423 non-null int16
dob_years           21423 non-null int8
education           21423 non-null object
education_id        21423 non-null int8
family_status       21423 non-null object
family_status_id    21423 non-null int8
gender              21423 non-null object
income_type         21423 non-null object
debt                21423 non-null int8
total_income        21423 non-null float32
purpose             21423 non-null object
years_employed      21423 non-null float64
dtypes: float32(1), float64(1), int16(1), int8(5), object(5)
memory usage: 1.4+ MB


### Conclusion

#### 
'days_employed' column has been converted to int type since we don't need such info as float.
The datatypes for the int and float types columns have been downcasted in order to save memory.

### Processing duplicates

In [13]:
#Removing the duplicated rows using the drop_duplicates() method.
df = df.drop_duplicates()
#making sure the duplicates are gone
df.duplicated().sum()

0

### Conclusion

#### 
Considering the amount of duplicated rows is very low compared to the entire dataset (54 out of 21525; 0.25%),
And the reason for the duplicated rows might be a human error, It shouldn't have an impact on our data and should be removed.


### Categorizing Data

#### Categorizing by children

In [14]:
#1. Categorazing by children
kids_relation = pd.pivot_table(df, index='children', columns='debt', values='family_status_id', aggfunc='count', margins=True).reset_index()
kids_relation

#changing NaN value to 0
kids_relation.fillna(0, inplace=True)

#creating a default rate by precentages
kids_relation['default_rate'] =(kids_relation[1]/kids_relation['All'])*100
kids_relation

#using the groupby method to show the same data
kids_relation2 = kids_relation.groupby(['children',0,1,'All','default_rate'])['children'].count()
kids_relation

debt children        0       1    All  default_rate
0           0  12979.0  1058.0  14037      7.537223
1           1   4398.0   442.0   4840      9.132231
2           2   1912.0   202.0   2114      9.555345
3           3    301.0    27.0    328      8.231707
4           4     37.0     4.0     41      9.756098
5           5      9.0     0.0      9      0.000000
6         All  19636.0  1733.0  21369      8.109879

#### Categorizing by martial status

In [15]:
#2. Categorizing by martial status
family_relation = pd.pivot_table(df, index='family_status', columns='debt', values='family_status_id', aggfunc='count', margins=True).reset_index()

#creating a default rate by precentages
family_relation['default_rate'] =(family_relation[1]/family_relation['All'])*100

#using the groupby method to show the same data
family_relation2 = family_relation.groupby(['family_status',0,1,'All','default_rate'])['family_status'].count()

family_relation

debt      family_status      0     1    All  default_rate
0     civil partnership   3755   386   4141      9.321420
1              divorced   1100    85   1185      7.172996
2               married  11368   927  12295      7.539650
3             unmarried   2521   273   2794      9.770938
4       widow / widower    892    62    954      6.498952
5                   All  19636  1733  21369      8.109879

#### categorizing by income level

In [16]:
#3. Categorizing by income level
df['total_income'].value_counts()

#realizing we have too much values in total_income, we need to categorize the income before we can check relations.
#We will create bins by income levels by the .describe method's output and .cut in order to create the  categorized column.

df['income_level'] = pd.cut(df['total_income'], bins=[0, 14233, 24149 , 31319, 100000, 362497], include_lowest=True,
       labels=['low', 'below_average', 'above_average', 'high','very_high'])

#Categorazing by income level
income_relation = pd.pivot_table(df, index='income_level', columns='debt', values='family_status_id', aggfunc='count', margins=True).reset_index()
income_relation['default_rate'] = (income_relation[1]/income_relation['All'])*100
income_relation

#using the groupby method to show the same data
income_relation2 = income_relation.groupby(['income_level',0,1,'All','default_rate'])['income_level'].count()
income_relation2

income_level   0      1     All    default_rate
All            19636  1733  21369  8.109879        1
above_average  3391   329   3720   8.844086        1
below_average  6362   603   6965   8.657574        1
high           4866   376   5242   7.172835        1
low            4924   419   5343   7.842036        1
very_high      93     6     99     6.060606        1
Name: income_level, dtype: int64

#### Categorizing by loan purpose

In [17]:
#4. Categorizing by loan purpose
df['purpose'].value_counts()

#we can see we have 38 values and many values with the same meaning, i.e : 'having a wedding' and 'to have a wedding'
#we will categorize them by meaning and rephrase the purpose column values with just the clear meaning to normalize the data.

#making another column purpose_type
df['purpose_type'] = df['purpose']


# rearranging the columns for better visualization
df = df[["children", "days_employed", "years_employed",'education','education_id','family_status','family_status_id',
'gender','income_type','debt','total_income','income_level','purpose','purpose_type']]

#mapping values to categories using dictionary in the 'purpose_type' column
search_map={
    'estate':['hous', 'proper', 'estate'],
    'wedding':['wedd'],
    'car':['car'],
    'education':['educ', 'univ']
    
}
for key, value in search_map.items():
    filter_column = df['purpose_type'].str.contains('|'.join(value))
    df['purpose_type'][filter_column] = key
    
    
# categorization by 4 groups has been done. creating pivot table to check for relations
purpose_relation = pd.pivot_table(df, index='purpose_type', columns='debt', values='family_status_id', aggfunc='count', margins=True).reset_index()

#creating a default rate by precentages
purpose_relation['default_rate'] =(purpose_relation[1]/purpose_relation['All'])*100
purpose_relation

#using the groupby method to show the same data
purpose_relation2 = purpose_relation.groupby(['purpose_type',0,1,'All','default_rate'])['purpose_type'].count()
purpose_relation2

/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:25: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy


purpose_type  0      1     All    default_rate
All           19636  1733  21369  8.109879        1
car           3886   400   4286   9.332711        1
education     3626   370   3996   9.259259        1
estate        9987   779   10766  7.235742        1
wedding       2137   184   2321   7.927617        1
Name: purpose_type, dtype: int64

### Conclusion

####  Categorization has been made by using pivot tables and by using groupby() method, by adding new category column and by a dictionary to map values into categories.

- Is there a relation between having kids and repaying a loan on time?

In [18]:
kids_relation

debt children        0       1    All  default_rate
0           0  12979.0  1058.0  14037      7.537223
1           1   4398.0   442.0   4840      9.132231
2           2   1912.0   202.0   2114      9.555345
3           3    301.0    27.0    328      8.231707
4           4     37.0     4.0     41      9.756098
5           5      9.0     0.0      9      0.000000
6         All  19636.0  1733.0  21369      8.109879

#### 
By making and viewing the pivot table that shows the relation between having kids and replaying loan on time, We can conclude several conclusions:
1. The average default rate is 8.1%.
2. Having 0 kids means having the lowest default rate besides 5 kids family - 7.53%.
3. Having 4 kids means having the highest default rate - 9.75%.
4. Having 3 kids means having the second lowest default rate - 8.23%
5. Families Having up to 4 kids have their default rate above average.
6. Families with 5 kids have default rate of 0% (total number of 9 families).



### Conclusion

#### 
There is a relation between having kids and replaying a loan on time , We can tell that having no kids means having the lowest default rate - highest chances to repay a loan on time.
On the other side, we can't correlate exactly by the children's number - families with 3 kids have a lower default rate than families with 1,2 or 4 children, and 9 families with 5 children have a default rate of 0 - but I would say 9 families is a too little of a number to conclude something.
To summerize, having no children at all means having the highest chance to repay on time, and having  up to 4 kids means having a lower chance to repay on time. 

- Is there a relation between marital status and repaying a loan on time?

In [19]:
family_relation

debt      family_status      0     1    All  default_rate
0     civil partnership   3755   386   4141      9.321420
1              divorced   1100    85   1185      7.172996
2               married  11368   927  12295      7.539650
3             unmarried   2521   273   2794      9.770938
4       widow / widower    892    62    954      6.498952
5                   All  19636  1733  21369      8.109879

#### By making and viewing the pivot table that shows the relation between marital status and replaying loan on time, We can conclude several conclusions:
1. The Average default rate is 8.1%.
2. Widows have the lowest default rate - 6.49%.
3. Unmarried have the highest default rate - 9.77%.
4. Being in a civil partnership means having a high default rate - 9.32%.

### Conclusion

#### There is a relation between marital status and repaying a loan on time.
We can conclude that being through marrige in all forms(still married, divorced and widowed) leads to having a lower default rate and a higher chance to repay on time.
the groups of married, divorced and widow all have lower default_rate than people who havent been through marrige - civil partnership and unmarried.

- Is there a relation between income level and repaying a loan on time?

In [20]:
income_relation

debt   income_level      0     1    All  default_rate
0               low   4924   419   5343      7.842036
1     below_average   6362   603   6965      8.657574
2     above_average   3391   329   3720      8.844086
3              high   4866   376   5242      7.172835
4         very_high     93     6     99      6.060606
5               All  19636  1733  21369      8.109879

#### By making and viewing the pivot table that shows the relation between marital status and replaying loan on time, We can conclude several conclusions:
1. The Average default rate is 8.1% .
2. very_high income level has the lowest default rate - 6.06%
3. below_average income level has the highest default rate - 8.65%
4. low income level has a lower default rate than below_average.

### Conclusion

#### There is a relation between income level and replaying a loan on time. 
We can conclude that people with high income (high and very_high income levels) have the lowest default rate highest chance to repay on time.
We can also conclude that the middle class (below_average and above_average) have the highest default rate and lowest chance to repay on time.
We also see that people with low income level have higher chance to repay their loan on time, perhaps due to the higher value and impact the loan is having on them compared to the middle class.

- How do different loan purposes affect on-time repayment of the loan?

In [21]:
purpose_relation

debt purpose_type      0     1    All  default_rate
0             car   3886   400   4286      9.332711
1       education   3626   370   3996      9.259259
2          estate   9987   779  10766      7.235742
3         wedding   2137   184   2321      7.927617
4             All  19636  1733  21369      8.109879

#### By making and viewing the pivot table that shows the relation between marital status and replaying loan on time, We can conclude several conclusions:
1. The average default rate is 8.1%.
2. The estate purpose of loan is having the lowest default_rate - 7.23%.
3. The car purpose of loan is having the highest default_rate - 9.33%.
4. Loan purpose of wedding is more likley to be paid on time than education purpose of loan.

### Conclusion

#### 
There is a relation between loan purpose and repaying a loan on time.
We can tell that loan purpose related to  estate is the most common (50.38%) and is most likely to be repaid on time (lowest default rate).
Loan purpose of car is having the highest default rate and is most likely to not be repaid on time.

## General conclusion

####
To summarize, we concluded many conclusions regarding different aspects of the data:

1. The data wasn't 'clean', it contained duplicates, missing values, and many illogical values.
2. We made the assumption that people that lack 'days_employed' and 'total_income' together aren't working and filled those        missing values with 0. The amount of those unemployed people are roughly 10% which makes sense (10% of the loan requests        made by unemployed people).
3. The duplicated rows have been removed (54 rows out of the entire data).
4. Illogical values (-1 children, 20 children, 0 years of age, 600 years of work etc) have been proccesed and are explained in    The 'Proccesing illogical values' block.
5. The integer types(int,float) data-types have been downcasted in order to save memory.

6. Categorization has been made according to the questions asked, pivot tables and tables using groupby method have been created for each categorization.

7. In matters of the data itself, We conclude that there are many relations in our data regarding paying the loan on time,
   And probably many more to discover in it.